### ISSUE TITLE
Add html reference to manual (as an online resource option as well)

### ISSUE THREAD

**Issue by porres** [pure-data/pddp#46]
I'll do this first => https://github.com/pure-data/pddp/issues/6

 later we can think of an online documentation reference

old discussion in https://github.com/pure-data/pddp/issues/10

----------------------------------------

Idea is to have all objects' basic description of functionality (example [float]: - store a float), then document messages to/from in/outles and creation arguments.

A first draft is here for the first 4 objects https://lucarda.github.io/pandoc-and-reference/

Pd-ceammc also has something similar, here for reference: https://ceammc.github.io/pd-help/help-en/

It could be part of Pd's manual as an appendix. This means it would live online in Miller's site and also be shipped with pd.

There could also be a dedicated link to it from within Pd (as in a link from Help menu).

**Comment by umlaeute** [comment]
it would be cool if you could flesh out an issue in a way, that other people understand what you mean by this.

"html reference" is very vague: i know it's written in HTML; but what exactly is *referenced* ´? just the objects? messages as well?...

this is true for many of these issues that only consist of a title (and probably some reference to another ticket; but really let's keep tickets understandable without external references )

**Comment by danomatika** [comment]
I would also suggest combining a number of small but related issues and put the related points together with a task list (aka TODO checkboxes). This lowers the amount of notification traffic which is already a lot for me.

See https://docs.github.com/en/issues/tracking-your-work-with-issues/about-task-lists

**Comment by Lucarda** [comment]
@porres see https://pd.iem.sh/list-of-objects/

only the first 4 object opens more description.
There might be a nicer layout in the `.md` of each description.

**Comment by porres** [comment]
Now that 0.52 is out and I made tons of changes/revisions it's time to start adding reference subpatches in the help files.

@Lucarda I don't really like this idea to index inlets/outlers from '0'. I'm using a standard.

here's my template so far

<img width="608" alt="Screen Shot 2021-12-20 at 01 03 58" src="https://user-images.githubusercontent.com/5360142/146710263-8ac8e385-b7a5-47c3-b188-52e5181571fa.png">

anyway, when done, we can copy these to an html

**Comment by porres** [comment]
the branch I'm adding reference subpatches -> https://github.com/porres/pure-data/tree/add-reference

**Comment by porres** [comment]
> It's time to work on the html files I guess, can I count on your help @Lucarda ?

this would require us to revise my initial work, which must still include mistakes and typos...

**Comment by umlaeute** [comment]
related: https://github.com/pure-data/pure-data/pull/1517#issuecomment-1056418148

**Comment by abreubacelar** [comment]
so maybe we could do the inverse, extract all the info porres put on the patches and create meta files and after create the mechanism that recreate the patches and html files from the meta files?

 so before anything... where is some description of how a meta file should be written? or we need to first agree on the format ourselves?

**Comment by Lucarda** [comment]
@porres

> can I count on your help @Lucarda ?

Yes

@abreubacelar

> so before anything... where is some description of how a meta file should be written? or we need to first agree on the format ourselves?

In pd-extended meta files where patches only containing pd comments. something like:

````
...
#N canvas 910 105 580 571 reference 0;
#X text 136 432 signal - signal output to sound card;
#X text 123 510 1) list - set output channels (default: 1 2);
#X text 132 160 signal - signal input from sound card;
...
````

This way we can easily parse the text. 
What is not yet clear to me is if this is possible in sub-patches (is there a way to know when a sub-patch ended?). May be there has to be one abstraction per reference? 

May be @umlaeute already has another idea?

**Comment by porres** [comment]
how would this work in help files that have the same help file? like [table], "garray", graphs and subpatches?

Or objects like text/array/file that have multiple functions?

For these cases I had different [pd reference] subpatches for each.

Also, some [pd reference] subpatches include multiple objects at once, like in the help file of fft~ and related objects...

I think that once this is done, it's not like we'll keep changing what the funcionalities are... and when something changes, we need to update the help file anyway... I'm usually the one doing that and I don't mind manually adding the same changes to the html files eventually...

**Comment by Lucarda** [comment]
@porres I would like to do some testing on how feasible is to auto-extract-and-deploy info from .pd files to end up, via HUGO, in the online resource (the now hypothetical https://pd.iem.sh).

To cut it short these are the needs for my proposed plan:

- (optional) a new branch on top of what you already had done.

- restrict the reference patches to only use comments.

- make a patch that extracts the info to a plain text file.

- do the stuff so that HUGO gets those text files into the HTML site.

We should start slowly with a couple of examples to evaluate all steps and the results.

We should give-up all the "visually niceness" both in .pd and .html files.

We should restructure the reference patches to fit or satisfy our new imposed needs.

We should do these tests together (you, me, @umlaeute , (others welcome)) so that we distribute our time, efforts and knowledge.

-------------------------

I would like to have green lights from you and @umlaeute .

If I don't get the green lights someone will have to do a big and slow copy/paste to get an html reference (also hard to maintain later).

**Comment by porres** [comment]
> We should give-up all the "visually niceness" both in .pd and .html files.

Doesn't seem worth it.

> someone will have to do a big and slow copy/paste to get an html reference

I can do it. Just help me out with it until I get the hang of it.

> (also hard to maintain later)

It's not like messages/methods change all of the time. I really think this is a one shot task that will require minimal headaches to keep it up. Anyway, I'm up for it...

**Comment by umlaeute** [comment]
Try again

**Comment by Lucarda** [comment]
ping @porres

I'm ready to start a "big and slow copy/paste".

Did you already started?

**Comment by porres** [comment]
not ready :/ I also couldn't clone the repository either yet...

**Comment by porres** [comment]
go ahead :)

**Comment by Lucarda** [comment]
@porres check https://pd.iem.sh/objects/. All objects in "General" were filled.

**Comment by porres** [comment]
great, let me have a look and get back to you

**Comment by Lucarda** [comment]
@porres all the "reference" patches that you did were transfered to the `.md`s

take a look at https://pd.iem.sh/objects/. You can also search for a given object via the "search bar".

Do you still have problems cloning `https://git.iem.at/pd/pd.iem.sh.git` ?

**Comment by porres** [comment]
> Do you still have problems cloning https://git.iem.at/pd/pd.iem.sh.git ?

yes :)

now please, explain me, you have some sort of automatic extraction? Cause I was going for a full revision first, then for a second revision while copying to the html

**Comment by porres** [comment]
> Do you still have problems cloning https://git.iem.at/pd/pd.iem.sh.git ?

yes :)

now please, explain me, you have some sort of automatic extraction? Cause I was going for a full revision first, then for a second revision while copying to the html

**Comment by porres** [comment]
did some quick checks and https://pd.iem.sh/objects/line/ has a wrong argument description, I wonder why, I checked the original that made into pd-0.52-2 and it is correct. By the way, I have already made some corrections in some revisions.

**Comment by Lucarda** [comment]
> you have some sort of automatic extraction?

no :) , that's why it took so long.

For me cloning (and pushing) worked on win11 and also on debian11. I can't tell whats going on for you on mac.

>  https://pd.iem.sh/objects/line/ has a wrong argument description, I wonder why, ...

some of the files were transfered prior pd-0.52-2. Anyway I think you should first be able to edit the files and familiarize with HUGO.

> I have already made some corrections in some revisions

The sooner you can clone/push the better.

**Comment by Lucarda** [comment]
Are you using https://git-scm.com/.

I'm using it on win and linux without problems.

**Comment by porres** [comment]
> For me cloning (and pushing) worked on win11 and also on debian11. I can't tell whats going on for you on mac.

I haven't tried really hard... I'll start by making a first round of revision, then a second one by copying to html

**Comment by Lucarda** [comment]
Ok.
Let me know when you start the revision on https://git.iem.at/pd/pd.iem.sh/
please create a new branch for it.

**Comment by porres** [comment]
well, this is a long discussion, just wanted to add that md files were written for vanilla objecst in plugdata. I will get them in this repository and we can compare to what we have and use a single revised source for both.

@Lucarda was the website at https://pd.iem.sh/objects/ generated with md files?

**Comment by Lucarda** [comment]
> @Lucarda was the website at https://pd.iem.sh/objects/ generated with md files?


@porres yes. I also think (here https://github.com/plugdata-team/plugdata/issues/75#issuecomment-1112514628) that you pointed to Plug-Data devs that the were already "reference" `.md` files at the https://git.iem.at/pd/pd.iem.sh/ repo.

why nobody likes https://pd.iem.sh/ ?

Is really a no "graphic/web designer" but fully functional. especially the "objects reference" that is searchable.

@umlaeute and I did all the framework so that you (or anyone else) continue getting content but that never happen. I wonder why?

**Comment by porres** [comment]
added files here: https://github.com/pure-data/pddp/tree/main/md-reference

**Comment by porres** [comment]
> why nobody likes https://pd.iem.sh/ ?

@timothyschoen  / @tomara-x  why don't you like it?

> Is really a no "graphic/web designer" but fully functional. especially the "objects reference" that is searchable.
> @umlaeute and I did all the framework so that you (or anyone else) continue getting content but that never happen. I wonder why?

I first wanna give yet another round of revision for the help files, then take care of these files revision too (both md files that you use and now plugdata as well, kinda cumbersome, I hope we can use the same source for all).

Where are the md files you're using anyway?

**Comment by Lucarda** [comment]
> Where are the md files you're using anyway?

https://git.iem.at/pd/pd.iem.sh/-/tree/main/content

**Comment by tomara-x** [comment]
hello hello! i didn't like it because i didn't know it existed? but not i do like it! hehe

**Comment by tomara-x** [comment]
oh and this feels like a "there are now 15 competing standards" situation. instead of making our own thing at plugdata, why not build on top of this, add our stuff, and link to it?

**Comment by porres** [comment]
> instead of making our own thing at plugdata, why not build on top of this, add our stuff, and link to it?

exactly :)

**Comment by porres** [comment]
Anyway, plugdata people also did md files for cyclone and else, I would love if these 2 could also be online somewhere for people to browse. Can we add folders or external libraries that anyone can go there and upload stuff?

**Comment by Lucarda** [comment]
> anyone can go there and upload stuff?

In theory yes. In practice I'm not sure if IEM's gitlab has an option to fork and made PRs. @umlaeute can tell us more. but of course we should be able to do that.

**Comment by porres** [comment]
> instead of making our own thing at plugdata, why not build on top of this, add our stuff, and link to it?

why not a github repo? Or a mirror or something?

**Comment by porres** [comment]
I know, but if it had a github mirror, we can make PRs to that which can then be merged into it...

**Comment by Lucarda** [comment]
sound good :)

Lets wait for some input from @umlaeute

**Comment by porres** [comment]
@Lucarda, I am finally touching this. Now, I'm not sure you're using the given .md files that I uploaded here to this repository. Here's a screenshot.

![Screen Shot 2023-03-12 at 15 18 52](https://user-images.githubusercontent.com/5360142/224564392-5a7c5a05-b12e-48e0-890f-6586e07453cf.png)

Note I am also editing and improving the reference subpatches, object descriptions and stuff. Here's my new reference for math functions:

![Screen Shot 2023-03-12 at 15 21 58](https://user-images.githubusercontent.com/5360142/224564510-fd443199-3813-4ea1-8e6e-484cd2ed00b6.png)

You see I now have a description for each object, so we don't have that of a vague description for all (namely "math function"). But then, the .md file I have here already gives a better description ("calculates absolute value").

The "see also" section is problematic, as in the help file we have grouped help files but in here we'd need to specify all objects and hopefully also group them like "binary operators" and stuff, I don't know, what do you think?

**Comment by Lucarda** [comment]
>  I'm not sure you're using the given .md files that I uploaded here to this repository.

no. I had not touched https://git.iem.at/pd/pd.iem.sh/-/tree/main/content since last year. I had not even looked at the files you uploaded. question to myself: are they compatible?


> The "see also" section is problematic, as in the help file we have grouped help files but in here we'd need to specify all objects and hopefully also group them like "binary operators" and stuff, I don't know, what do you think?

I think for now on the `.md` files we just put each objects. not categories. also my head has been out of this for long and i'll have re dive on it (HUGO and all that).

also @umlaeute has happily ignored questions about github mirrors and stuff. Not sure what to do with all this.

**Comment by porres** [comment]
> I had not even looked at the files you uploaded. question to myself: are they compatible?

I suspect may have gotten files that are not the ones that are being used somehow, because of the discrepancy I pointed... let's follow through this, I wanna be sure I have the right ones that are currently used to load the existing/accessible html files. Let me first work on it and then we see how to upload, make a PR, etc...

**Comment by umlaeute** [comment]
> also @umlaeute has happily ignored questions about github mirrors and stuff. Not sure what to do with all this.

git is a decentralized VCS. You can mirror it wherever you want. There currently is no license attached to the data in the repository (which we should change), but I assume we want this to be open-source (BSD, or AGPL) so you can even put it into a public repository of your liking.

**Comment by Lucarda** [comment]
Ok just some questions:

- can we mirror it here https://github.com/pure-data/pddp/ so people can clone and make PRs?

- when the PRs got merged the site at IEM rebuilds?

**Comment by porres** [comment]
we'd mirror it somewhere else, as a mirror, not into pddp... or we can have the .md files here and the gitlab repo could use it if it becomes a submodule (another idea)

**Comment by porres** [comment]
> no. I had not touched https://git.iem.at/pd/pd.iem.sh/-/tree/main/content since last year. I had not even looked at the files you uploaded. question to myself: are they compatible?

@Lucarda, the files I uploaded here are the ones found in this link/repository, so I do find it very strange that the actual information from the files that we can fin in https://pd.iem.sh/list-of-objects/ is different, like I mentioned in https://github.com/pure-data/pddp/issues/46#issuecomment-1465267601

maybe @umlaeute can helps us sort it out. But I'm gonna just work on these files we have as I should revise and redo things anyway...

I'll come back with the revised files when 0.54-0 is out and my changes in the documentation branch have been merged

**Comment by umlaeute** [comment]
> and the gitlab repo could use it if it becomes a submodule (another idea)

i don't like this very much (as it appears that this repo is about multiple Pd-flavours; and in general i heartily dislike `git submodule`; for one it complicates the mirroring and automatic pulling).
just my 2¢

**Comment by porres** [comment]
> it appears that this repo is about multiple Pd-flavours

but it shouldn't really, I know I also included plugdata's md files, but the idea was to have the same infor for both (which I guess can't happen)

**Comment by porres** [comment]
so, @Lucarda why do we have "methods" separately instead of messages under the "1st" inlet?

and does it matter the order things are? Like, why arguments usually come first, can't I just follow the order of the actual help files/reference?

**Comment by Lucarda** [comment]
> why do we have "methods" separately instead of messages under the "1st" inlet?

I don't know, may be that was how the pd-reference help patches were made a year ago.

> and does it matter the order things are? Like, why arguments usually come first, can't I just follow the order of the actual help files/reference?

you can change the order in the `.md` files as you wish, note that the end result (the displayed html) follows the order set in https://git.iem.at/pd/pd.iem.sh/-/blob/main/layouts/partials/object-reference.html but this can also be changed.

if you change the order in the `.md` files be careful with the indentation. You can always check by running `hugo server`: if something went wrong it will break the rendering and tell you where is the error.

**Comment by umlaeute** [comment]
>  why do we have "methods" separately instead of messages under the "1st" inlet?

where do we have them?
i see exactly 3 occurences (in https://git.iem.at/pd/pd.iem.sh/-/tree/main/content/objects/)
- *file-handle*
- `*ile* (which is practically a copy of *file-handle*)
- `*array*

i'm not exactly sure why there are `methods` in *file-handle* (although the "methods" section just covers in-detail info on some messages for the 1st inlet).
as for *garray*, that object doesn't have any inlets. and looking at the `canvas-help.pd` it says "methods". so the only one to answer this question is @porres :-)

**Comment by porres** [comment]
sorry, I was looking at plugdata's files :( let me just remove them to avoid more screw ups :)

**Comment by timothyschoen** [comment]
> so, @Lucarda why do we have "methods" separately instead of messages under the "1st" inlet?
> 
> and does it matter the order things are? Like, why arguments usually come first, can't I just follow the order of the actual help files/reference?

This is a change to the format we made for plugdata, since we use these docs to provide tooltips. Putting all the methods inside the tooltip would be too much. (porres already knew that but in case anyone else is wondering)

<img width="299" alt="Screenshot 2023-03-21 at 17 34 42" src="https://user-images.githubusercontent.com/44585538/226678790-7357347a-d0a7-43f3-a342-6688c8419418.png">

**Comment by porres** [comment]
@timothyschoen i guess you can have your special .md files for tooltips, but then we just use our own for the different purpose of working on an html file that you can also incorporate into plugdata when it's ready :) so I'll just remove you md files from this repository

**Comment by porres** [comment]
@Lucarda what's the easy and best way for me to check the changes in html format? please give me baby steps

**Comment by Lucarda** [comment]
### hugo

install `hugo`

https://gohugo.io/installation/macos/#package-managers

then clone https://git.iem.at/pd/pd.iem.sh (you can sign-in with your github account)

on a terminal placed on the cloned folder do `hugo server`

you will see something like this on the terminal:

```
lucarda@t410w11 MINGW64 /f/git-portable-2/pd.iem.sh
$ hugo server
Start building sites …
hugo v0.109.0-7238115e146b21ecbd872ef1c0b08eb7f282fd25 windows/amd64 BuildDate=2022-12-25T17:00:25Z

                   | EN
-------------------+------
  Pages            | 286
  Paginator pages  |   0
  Non-page files   |   0
  Static files     |  12
  Processed images |   0
  Aliases          |  20
  Sitemaps         |   1
  Cleaned          |   0

Built in 701 ms
Watching for changes in F:\git-portable-2\pd.iem.sh\{archetypes,content,layouts,static}
Watching for config changes in F:\git-portable-2\pd.iem.sh\config.toml
Environment: "development"
Serving pages from memory
Running in Fast Render Mode. For full rebuilds on change: hugo server --disableFastRender
Web Server is available at //localhost:1313/ (bind address 127.0.0.1)
Press Ctrl+C to stop

```

go to `safari` and put this in the address bar `localhost:1313` hit enter.

now you have a live preview.

edit the `.md` files in `pd.iem.sh/content/objects`. when you hit save in the `.md` file the web browser gives a preview (just browse to the object you are editing).

on the terminal Press Ctrl+C to stop hugo when you are done or you need to restart.

### places of the `.md` files

so it seems you are having copies of the objects `.md` files. I think it's best if you keep the work in https://git.iem.at/pd/pd.iem.sh (you can sign-in with your github account). you can create a branch. If not just keep a backup of `pd.iem.sh/content/objects` somewhere in your Pc.

**Comment by porres** [comment]
thanks for all the steps

I'm getting a weird error trying to fork "An error occurred while forking the project. Please try again."

**Comment by Lucarda** [comment]
you can't fork this.
can you create a branch on iem's gitlab?
if not we ask @umlaeute so you have write access.

**Comment by porres** [comment]
I created https://git.iem.at/pd/pd.iem.sh/-/tree/porres-revision what do you mean then by me 'cloning' it? How do I do it? I know... stupid question... but I just always used github desktop and could use some help while you're here :)

**Comment by porres** [comment]
trying to clone with github desktop gave me this, seems like I need some password

![Screen Shot 2023-03-22 at 00 04 34](https://user-images.githubusercontent.com/5360142/226791486-334e981b-7e4c-40e6-ae43-2c73cfcf4a72.png)

**Comment by porres** [comment]
I guess there are "terminal ways" to do this, but I'd really love to have github's IDE on this process

**Comment by Lucarda** [comment]
try to clone the https: `https://git.iem.at/pd/pd.iem.sh.git`

?

**Comment by Lucarda** [comment]
seems you were using the "Clone with SSH"

**Comment by porres** [comment]
tried this

![Screen Shot 2023-03-22 at 00 09 49](https://user-images.githubusercontent.com/5360142/226792161-de677695-9c29-4235-97f7-db3203bb8fea.png)

got this

![Screen Shot 2023-03-22 at 00 09 14](https://user-images.githubusercontent.com/5360142/226792188-bbe785fd-58f7-46ef-addf-df067246568f.png)

**Comment by Lucarda** [comment]
wait. I'll try on my Pc ...

**Comment by Lucarda** [comment]
works for me:

```
$ git clone https://git.iem.at/pd/pd.iem.sh.git
Cloning into 'pd.iem.sh'...
remote: Enumerating objects: 3981, done.
remote: Counting objects: 100% (3813/3813), done.
remote: Compressing objects: 100% (995/995), done.
remote: Total 3981 (delta 2736), reused 3788 (delta 2711), pack-reused 168
Receiving objects: 100% (3981/3981), 2.65 MiB | 1.44 MiB/s, done.
Resolving deltas: 100% (2802/2802), done.
```

**Comment by Lucarda** [comment]
how about if you open a terminal on your desktop or other dir and do

`git clone https://git.iem.at/pd/pd.iem.sh.git`

**Comment by Lucarda** [comment]
are you using the correct time date year on your machine?

**Comment by Lucarda** [comment]
try to delete certificate as in:

https://superuser.com/questions/721778/can-not-clone-any-git-repository-from-github

?

**Comment by Lucarda** [comment]
or upgrade git ?

**Comment by Lucarda** [comment]
there must be some "Expired Certificates" on your mac

**Comment by porres** [comment]
> there must be some "Expired Certificates" on your mac

nope :/

**Comment by porres** [comment]
it'd be really great for me if there was some github mirror of this... I remember I tried all this like a year ago and also ran into troubles and gave up

**Comment by porres** [comment]
> nope :/
![Screen Shot 2023-03-22 at 01 40 58](https://user-images.githubusercontent.com/5360142/226803865-31d122e6-e894-45d2-9528-e02302774f7f.png)

**Comment by Lucarda** [comment]
i think you did

1. Open Keychain Access
1. search Certificates for "DigiCert"
1. View menu > Show Expired Certificates


what happens if you do

1. Open Keychain Access
1. View menu > Show Expired Certificates

**Comment by Lucarda** [comment]
just for the record, I did

`git clone https://git.iem.at/pd/pd.iem.sh.git`

on my Monterrey VM and it worked fine.

**Comment by porres** [comment]
> what happens if you do

that's what I had done

![Screen Shot 2023-03-22 at 10 07 23](https://user-images.githubusercontent.com/5360142/226913966-7e926241-0c6a-41cb-8567-745904b5c31a.png)

**Comment by Lucarda** [comment]
from your pic above, see yellow:

![certs](https://user-images.githubusercontent.com/25239522/226915890-b07b7d5c-9649-41be-8a12-0c657480e50c.png)

**Comment by porres** [comment]
dude, thanks for helping me through, but from the start, I know what I was doing. I'll just say there are NO expired certificates... no screenshots no more, just trust me, screenshots were supposed to prove and settle it, didn't work :)

> you have "digi" in the search. there should be a lot more certs

this was the one people mentioned in https://github.com/pure-data/pddp/issues/46#issuecomment-1478873847 I had cleaned things up, still can't clone in github desktop, will try some more, install or update git via terminal or someting, redownload it, but if in the end I just get to clone via terminal, I won't be happy and I could just download source and do this instead...

**Comment by porres** [comment]
it'd be great though if there was a github mirror to make things easier

**Comment by Lucarda** [comment]
> it'd be great though if there was a github mirror to make things easier

ok, i'm goining to be AWFK for 10hrs. I agree that some github mirror or repo will be easier. let's wait for some @umlaeute input.

**Comment by porres** [comment]
yay, thanks for the github mirror

**Comment by umlaeute** [comment]
guys: it would be great if could move chat-like conversations that have nothing to do with the actual issue (just for reference, it is "Add html reference to manual (as an online resource option as well)") to "some other place".
e.g. a dedicated ticket (e.g. on https://git.iem.at/pd/pd.iem.sh/-/issues) or a real chat on IRC/telegram/discord/...

**Comment by Lucarda** [comment]
> This is a change to the format we made for plugdata, since we use these docs to provide tooltips. Putting all the methods inside the tooltip would be too much. (porres already knew that but in case anyone else is wondering)

@timothyschoen so IIUC these information you added is only on UI objects? (the ones  https://pd.iem.sh does not cover yet).

there are lots of chances that we all benefit from using https://github.com/pd-projects/hugo-puredata.info as the central place for objects .md files.

**Comment by porres** [comment]
> it would be great if could move chat-like conversations (...) to "some other place".

I'm also annoyed about this issue being a chat channel, so here we go, we can keep it at ---> https://github.com/pure-data/pddp/discussions/141
